### PR TITLE
add ripple ledger

### PR DESCRIPTION
**Pull request by siovanus**
add ripple ledger
1. update side chain manager
2. add ripple handler

### CODE REVIEW

File: native/service/governance/side_chain_manager/side_chain_manager.go [hunk 1]
```diff
@@ -431,3 +447,122 @@ func SetBtcTxParam(native *native.NativeService) ([]byte, error) {
...
+	if err != nil {
+		return utils.BYTE_FALSE, fmt.Errorf("RegisterAssetMap, checkWitness error: %v", err)
+	}
+
+	operatorAddress, err := GetOperatorAddress(native)
+	if err != nil {
+		return utils.BYTE_FALSE, fmt.Errorf("RegisterAssetMap, GetOperatorAddress error: %v", err)
+	}
+	if operatorAddress == common.ADDRESS_EMPTY {
+		PutOperatorAddress(native, params.OperatorAddress)
```
Comment (devfans) [inline native/service/governance/side_chain_manager/side_chain_manager.go:468]: Operator may better to be set when side chain is registered, and isolated with chainIDs

File: native/service/cross_chain_manager/ripple/ripple_handler.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+		signature, err := hex.DecodeString(s.Signer.TxnSignature)
+		if err != nil {
+			return fmt.Errorf("MultiSign, hex.DecodeString signature error: %s", err)
+		}
+
+		// check if valid signer
+		flag := false
+		for _, v := range rippleExtraInfo.Pks {
+			if fmt.Sprintf("%X", v) == s.Signer.SigningPubKey {
+				flag = true
```
Comment (devfans) [inline native/service/cross_chain_manager/ripple/ripple_handler.go:94]: break missing? convert s.Signer.SigningPubKey to same format?

File: native/service/cross_chain_manager/ripple/ripple_handler.go [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+			return fmt.Errorf("MultiSign, types.CheckMultiSign error: %s", err)
+		}
+		signer := &Signer{
+			Account:       signerAccount.Bytes(),
+			TxnSignature:  signature,
+			SigningPubKey: signerPk,
+		}
+		sink := common.NewZeroCopySink(nil)
+		signer.Serialization(sink)
+		multisignInfo.SigMap[hex.EncodeToString(sink.Bytes())] = true
```
Comment (devfans) [inline native/service/cross_chain_manager/ripple/ripple_handler.go:113]: Possible multiple signatures with same account?

File: native/service/governance/side_chain_manager/side_chain_manager.go [hunk 4]
```diff
@@ -50,6 +57,12 @@ const (
 	BIND_SIGN_INFO            = "bindSignInfo"
 	BTC_TX_PARAM              = "btcTxParam"
 	REDEEM_SCRIPT             = "redeemScript"
+	ASSET_MAP                 = "assetMap"
+	ASSET_MAP_INDEX           = "assetMapIndex"
+	RIPPLE_EXTRA_INFO         = "rippleExtraInfo"
+	FEE                       = "fee"
+	FEE_INFO                  = "feeInfo"
+	OPERATOR_ADDRESS          = "operatorAddress"
```
[the remark this answers is not in the document]
Comment (siovanus) [reply native/service/governance/side_chain_manager/side_chain_manager.go:65]: It's not only used for info but for ripple like chains, the fee storage have chain id prefix so if we have a chain like btc, we can use this method too.

File: native/service/cross_chain_manager/ripple/ripple_handler.go [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+		return fmt.Errorf("ripple MakeTransaction, get asset map error: %s", err)
+	}
+	assetAddress := assetMap.AssetMap[param.ToChainID]
+
+	// get rippleExtraInfo
+	rippleExtraInfo, err := side_chain_manager.GetRippleExtraInfo(service, param.ToChainID, assetAddress)
+	if err != nil {
+		return fmt.Errorf("ripple MakeTransaction, side_chain_manager.GetRippleExtraInfo error")
+	}
+
```
[the remark this answers is not in the document]
Comment (siovanus) [reply native/service/cross_chain_manager/ripple/ripple_handler.go:190]: done by return error if rippleExtraInfo is empty

File: native/service/cross_chain_manager/ripple/ripple_handler.go [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+	//get fee
+	baseFee, err := side_chain_manager.GetFee(service, param.ToChainID)
+	if err != nil {
+		return fmt.Errorf("ripple MakeTransaction, side_chain_manager.GetFee error: %v", err)
+	}
+	if baseFee.View == 0 {
+		return fmt.Errorf("ripple MakeTransaction, base fee is not initialized")
+	}
```
[the remark this answers is not in the document]
Comment (siovanus) [reply native/service/cross_chain_manager/ripple/ripple_handler.go:198]: Here this error means baseFee is empty, if baseFee is empty, baseFee.View is 0 and base fee is not initialized.

File: native/service/governance/side_chain_manager/side_chain_manager.go (hunk 1 above)
Comment (siovanus) [reply native/service/governance/side_chain_manager/side_chain_manager.go:468]: That's a good idea

File: native/service/cross_chain_manager/ripple/ripple_handler.go (hunk 2 above)
Comment (siovanus) [reply native/service/cross_chain_manager/ripple/ripple_handler.go:94]: Here check s.Signer.SigningPubKey is the specific multi signer's publickey, so convert s.Signer.SigningPubKey to same format and check if s.Signer.SigningPubKey is in multi signer group.

File: native/service/cross_chain_manager/ripple/ripple_handler.go (hunk 3 above)
Comment (siovanus) [reply native/service/cross_chain_manager/ripple/ripple_handler.go:113]: If multiple signatures with same account and same raw msg, the signature after signed is the same too, so the map will be covered and length count do not +1

File: native/service/cross_chain_manager/ripple/ripple_handler.go [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+		if err := scom.PutDoneTx(service, txParam.CrossChainID, params.SourceChainID); err != nil {
+			return nil, fmt.Errorf("vote MakeDepositProposal, PutDoneTx error:%s", err)
+		}
+
+		//fulfill to contract address
+		assetBind, err := side_chain_manager.GetAssetBind(service, params.SourceChainID)
+		if err != nil {
+			return nil, fmt.Errorf("vote MakeDepositProposal, side_chain_manager.GetAssetBind error:%s", err)
+		}
+		txParam.ToContractAddress = assetBind.LockProxyMap[txParam.ToChainID]
```
Comment (devfans) [inline native/service/cross_chain_manager/ripple/ripple_handler.go:95]: Check bind proxy empty here?

File: native/service/cross_chain_manager/ripple/ripple_handler.go [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+		dstAddress, eof := source.NextVarBytes()
+		if eof {
+			return nil, fmt.Errorf("vote MakeDepositProposal, deserilize dst address error:%s", err)
+		}
+		amount, eof := source.NextUint64()
+		if eof {
+			return nil, fmt.Errorf("vote MakeDepositProposal, deserilize amount error:%s", err)
+		}
+		s := common.NewZeroCopySink(nil)
+		s.WriteVarBytes(assetBind.AssetMap[txParam.ToChainID])
```
Comment (devfans) [inline native/service/cross_chain_manager/ripple/ripple_handler.go:108]: Also check empty bind asset here

**Review by devfans — APPROVED** [review]

**Pull request merged**